flang/lib/Lower/CallInterface.cpp

```diff
@@ -531,6 +543,17 @@ class Fortran::lower::CallInterfaceImpl {
     }
   }
 
+  void appendExtraArgument(mlir::Type tupTy) {
+    auto *ctxt = tupTy.getContext();
+    addFirOperand(tupTy, nextPassedArgPosition(), Property::BaseAddress,
+                  {mlir::NamedAttribute{
+                      mlir::Identifier::get(fir::getHostAssocAttrName(), ctxt),
+                      mlir::UnitAttr::get(ctxt)}});
+    interface.passedArguments.emplace_back(
+        PassedEntity{PassEntityBy::BaseAddress, std::nullopt,
+                     interface.side().getHostAssociatedTuple(), emptyValue()});
```

Comment:
I do not think you need to create a `PassedEntity` for the host tuple argument. The idea of making a difference between FIR ("FirOperand") and Fortran arguments ("PassedEntity" that are symols or expression) was that their could be more "FirOperand" than "PassedEntities".

So I think you could very well keep the `PassedEntity` type as it was before, and not add a PassedEntity here.

`CalleeInterface::getHostAssociatedTuple()` could be the one looking for the mlir::Value in the FuncOp (like it is done [here ](https://github.com/flang-compiler/f18-llvm-project/blob/01cf5894e00a32cc1906e0e53c96de34fe4f7d02/flang/lib/Lower/HostAssociations.cpp#L65) in your PR, or simply assuming the last position an verifying the argument attribute).

Then, in the Bridge, you could have:
```
    for (const auto &arg : callee.getPassedArguments())
      // ....
    }
    if (auto passedResult = callee.getPassedResult()) {
       // ....
    }
   if (funit.parentHasHostAssoc())
     funit.parentHostAssoc().internalProcedureBindings(
              *this, localSymbols, funit.getHostAssoc(), callee.getHostAssociatedTuple());
```

It's all functionally equivalent, but in my opinion would keep internal procedure handling more orthogonal to normal argument handling.

Reply:
> I do not think you need to create a PassedEntity for the host tuple argument. The idea of making a difference between FIR ("FirOperand") and Fortran arguments ("PassedEntity" that are symols or expression) was that their could be more "FirOperand" than "PassedEntities".

Yes, this was what I was thinking. There is no operand at all in Fortran, and we're lambda lifting the host associations as FIR operands, but the associations one needs from this context seem backward or, at least, painful.

I'll see if I can't restructure this as you suggest. I wasn't fond of trying to tweak in a fake argument here.

Reply:
Looking more at the CallInterface, I see that `mapPassedEntities` is expecting that there is PassedEntity somewhere for every fir arguments (excepts for fir arguments that are Fortran results).
So one way could be to have a special `FirPlaceHolder::passedEntityPosition` code for the host tuple like `FirPlaceHolder::resultEntityPosition`.
But I anyway think the CallInterface will need a bit of simplification at some point, so I am OK keeping things like you did.